src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +327,279 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
+  CpaCySymSessionSetupData sessionSetupData;
+  memset(&sessionSetupData, 0, sizeof(sessionSetupData));
+
+  sessionSetupData.sessionPriority = CPA_CY_PRIORITY_NORMAL;
+  sessionSetupData.symOperation = CPA_CY_SYM_OP_CIPHER;
+  sessionSetupData.cipherSetupData.cipherAlgorithm = CPA_CY_SYM_CIPHER_AES_CBC;
+  sessionSetupData.cipherSetupData.cipherKeyLenInBytes = AES_256_KEY_SIZE;
+  sessionSetupData.cipherSetupData.pCipherKey = pCipherKey;
+  sessionSetupData.cipherSetupData.cipherDirection = cipherDirection;
+
+  status = cpaCySymDpSessionCtxGetSize(cyInstHandle, &sessionSetupData, &sessionCtxSize);
+  if (CPA_STATUS_SUCCESS == status) {
+    status = qcc_contig_mem_alloc((void **)(sessionCtx), sessionCtxSize);
+  } else {
+    dout(1) << "cpaCySymDpSessionCtxGetSize failed with status = " << status << dendl;
   }
-  if (qcc_inst->is_polled[avail_inst] == CPA_TRUE) {
-    while (!qcc_op_mem[avail_inst].op_complete) {
-      icp_sal_CyPollInstance(qcc_inst->cy_inst_handles[avail_inst], 0);
+  if (CPA_STATUS_SUCCESS == status) {
+    do {
+      status = cpaCySymDpInitSession(cyInstHandle,
+                                  &sessionSetupData,
+                                  *sessionCtx);
+    } while (CPA_STATUS_RETRY == status);
+    if (status != CPA_STATUS_SUCCESS) {
+      dout(1) << "cpaCySymDpInitSession failed with status = " << status << dendl;
     }
+  } else {
+    dout(1) << "Session alloc failed with status = " << status << dendl;
   }
-  pthread_join(cypollthreads[avail_inst], NULL);
+  return status;
+}
 
-  if(qcc_op_mem[avail_inst].op_result != CPA_STATUS_SUCCESS) {
-    derr << "Unable to perform crypt operation" << dendl;
-    return false;
-  }
+bool QccCrypto::symPerformOp(int avail_inst,
+                              CpaCySymSessionCtx sessionCtx,
+                              const Cpa8U *pSrc,
+                              Cpa8U *pDst,
+                              Cpa32U chunk_size,
+                              Cpa32U total_len,
+                              Cpa8U *pIv,
+                              Cpa32U ivLen) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  Cpa32U one_batch_size = chunk_size * MAX_NUM_SYM_REQ_BATCH;
+  for (Cpa32U off = 0; off < total_len; off += one_batch_size) {
+    CompletionHandle complete;
+    for (Cpa32U offset = off, i = 0; offset < total_len && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
+      CpaCySymDpOpData *pOpData = qcc_op_mem[avail_inst].sym_op_data[i];
+      Cpa8U *pSrcBuffer = qcc_op_mem[avail_inst].src_buff[i];
+      Cpa8U *pIvBuffer = qcc_op_mem[avail_inst].iv_buff[i];
+      Cpa32U process_size = offset + chunk_size <= total_len ? chunk_size : total_len - offset;
+      if (CPA_STATUS_SUCCESS == status) {
+        // copy source into buffer
+        memcpy(pSrcBuffer, pSrc + offset, process_size);
+        // copy IV into buffer
+        memcpy(pIvBuffer, &pIv[i * ivLen], ivLen);
+      }
+      complete.add_one();
+      if (CPA_STATUS_SUCCESS == status) {
```

Comment:
this can remove and merge into the above if (CPA_STATUS_SUCCESS == status)

Reply:
done.